app/ts/utils/eip712.ts

```diff
@@ -0,0 +1,398 @@
+	// Check for tuple types e.g. tuple(uint256,MyStruct) or tuple(uint256,MyStruct)[]
+	if (typeStr.startsWith('tuple(')) return { valid: false, reason: 'tuples are not supported'}
+
+	// Check for array types (non-tuple arrays)
+	const arrayMatch = typeStr.match(/^([^\[]+)(\[(?!0\])[0-9]*\](\[(?!0\])[0-9]*\])*)$/)
+	if (arrayMatch) {
+		const base = arrayMatch[1] // "array"
+		const brackets = arrayMatch[2] // "[][][]"
+		if (base === undefined || brackets === undefined) return { valid: false, reason: 'base or brackets were undefined'}
+		const remaining = brackets.replace(/^(\[(?!0\])[0-9]*\])/, '')
+		const innerType = base + remaining
+		const arrayLength = (brackets: string | undefined) => {
+			if (brackets === undefined) return undefined
+			const lengthMatch = brackets.match(/\[(\d*)\]$/)
+			return lengthMatch && lengthMatch[1] !== '' && lengthMatch[1] !== undefined ? parseInt(lengthMatch[1]) : undefined
+		}
+		const expectedLength = arrayLength(arrayMatch[2])
+		if (!Array.isArray(value)) return { valid: false, reason: `invalid array: ${ JSON.stringify(value) }` }
+
+		if (expectedLength !== undefined && value.length > expectedLength) return { valid: false, reason: `expected array of length ${ expectedLength }, got ${ value.length }` }
+		for (const elem of value) {
+			const valid = validateTypeValue(innerType, elem, solidityTypeTree)
+			if (valid.valid === false) return valid
+		}
+		return { valid: true }
+	}
+	// Now typeStr is expected to be a primitive or custom type
+	return validatePrimitiveOrStruct(typeStr, value, solidityTypeTree)
+}
+
+const getBaseType = (typeStr: string): string => typeStr.match(/^([^\[]+)/)?.[1] ?? typeStr
+
+const validatePrimitiveOrStruct = (typeStr: string, value: typeJSONEncodeable, solidityTypeTree: SolidityTypeTree): { valid: true } | { valid: false, reason: string } => {
+	const identifierPattern = /^[a-zA-Z_][a-zA-Z0-9_]*$/
+	if (!identifierPattern.test(typeStr)) return { valid: false, reason: `invalid type: ${ typeStr }` }
+
+	// Check for integer types, e.g. uint128 or int256
+	const intRegex = /^(u?)int(\d+)$/
+	const intMatch = typeStr.match(intRegex)
+	if (intMatch) {
+		// Use BigInt to check numeric ranges
+		const isUnsigned = intMatch[1] === 'u'
+		if (!intMatch[2]) return { valid: false, reason: `intMatch was undefined` }
+		const bitWidth = parseInt(intMatch[2])
+		// For unsigned: range is 0 to 2^width - 1
+		// For signed: range is -2^(width-1) to 2^(width-1) - 1
+		const min = isUnsigned ? 0n : -(2n ** BigInt(bitWidth - 1))
+		const max = isUnsigned ? (2n ** BigInt(bitWidth)) - 1n : (2n ** BigInt(bitWidth - 1)) - 1n
+
+		let bigValue: bigint
+		if (typeof value === 'number') {
+			if (!Number.isInteger(value)) return { valid: false, reason: `${ JSON.stringify(value) } wasn't integer` }
+			bigValue = BigInt(value)
+		} else if (typeof value === 'string') {
+			const parsed = Eip712Number.safeParse(value) // todo, does the quantity need to be in hex format?
```

Comment:
Does this `TODO` need to be addressed?  Are there any others?

Reply:
The spec doesn't specify this but in practical cases the numbers can be represented in many formats. So I am assuming its correc